vignettes/loading-data.Rmd

```diff
@@ -0,0 +1,198 @@
+---
+title: "Loading data"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Creating tensors}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r setup}
+#library(torch)
+```
+
+## Datasets and data loaders
+
+Central to data ingestion and preprocessing are datasets and data loaders.
+
+`torch` comes equipped with a bag of datasets related to, mostly, image recognition and natural language processing (e.g.,
+`mnist_dataset()`), which can be iterated over by means of `dataloader`s:
+
+```{r, eval = FALSE}
+# ...
+ds <- mnist_dataset(
+  dir, 
+  download = TRUE, 
+  transform = function(x) {
+    x <- x$to(dtype = torch_float())/256
+    x[newaxis,..]
+  }
+)
+
+dl <- dataloader(ds, batch_size = 32, shuffle = TRUE)
+
+for (b in enumerate(dl)) {
+  # ...
+
+```
+
+Cf. `vignettes/examples/mnist-cnn.R` for a complete example.
+
+What if you want to train on a different dataset? In these cases, you subclass `Dataset`, an abstract container that needs to
+know how to iterate over the given data. To that purpose, your subclass needs to implement `.getitem()`, and say what should
+be returned when the data loader is asking for the next batch.
+
+In `.getitem()`, you can implement whatever preprocessing you require. Additionally, you should implement `.length()`, so
+users can find out how many items there are in the dataset.
+
+While this may sound complicated, it is not at all. The base logic is straightforward -- complexity will, naturally, correlate
+with how involved your preprocessing is. To provide you with a simple but functional prototype, here we show how to create
+your own `dataset` to train on [Allison Horst\'s penguins](https://github.com/allisonhorst/palmerpenguins).
+
+## A custom dataset
+
+```{r}
+# remotes::install_github("allisonhorst/palmerpenguins")
+library(palmerpenguins)
```

Comment:
I think we need to add palmerpenguins to the Suggests in the DESCRIPTION file
If it's only on github we also need the Remotes field

Same for `dplyr` and `purrr`. (Perhaps it's better to avoid them to not increase the CI testing time)

Reply:
yeah, I see... removing `purrr` is also what you suggest below  - and removing `dplyr` isn't that hard either so I'll do that :-)

Reply:
hm, for palmerpenguins ... this: https://stackoverflow.com/questions/36105257/how-to-make-r-package-recommend-a-package-hosted-on-github sounds pretty awkward ... and it seems like they're in the process of putting it on CRAN ... https://github.com/allisonhorst/palmerpenguins/commit/8e1061dc0b3d0cbb53bea0545f651950b2109dca

perhaps i just add it to Suggests for now?